wallet/src/account/mod.rs

```diff
@@ -0,0 +1,407 @@
+                }
+                Destination::PublicKey(pk) => {
+                    self.key_chain.mark_public_key_as_used(db_tx, pk)?;
+                }
+                Destination::AnyoneCanSpend
+                | Destination::ClassicMultisig(_)
+                | Destination::ScriptHash(_) => {}
+            }
+        }
+        Ok(())
+    }
+
+    pub fn get_balance<B: Backend>(
+        &self,
+        db_tx: &StoreTxRo<B>,
+    ) -> WalletResult<(Amount, BTreeMap<TokenId, Amount>)> {
+        let utxos = db_tx.get_utxo_set(&self.get_account_id())?;
+        let balances = Self::calculate_output_amounts(utxos.values().map(|utxo| utxo.output()))?;
+        Ok(balances)
+    }
+
+    pub fn reset_to_height<B: Backend>(
+        &mut self,
+        db_tx: &mut StoreTxRw<B>,
+        block_height: BlockHeight,
+    ) -> WalletResult<()> {
+        let revoked_utxos: Vec<AccountOutPointId> = db_tx
+            .get_utxo_set(&self.get_account_id())?
+            .into_iter()
+            .filter_map(|(utxo_id, utxo)| match utxo.source() {
+                UtxoSource::Blockchain(utxo_height) if *utxo_height >= block_height => {
+                    Some(utxo_id)
+                }
+                UtxoSource::Mempool | UtxoSource::Blockchain(_) => None,
+            })
+            .collect();
+
+        for utxo_id in revoked_utxos {
+            db_tx.del_utxo(&utxo_id)?;
+        }
+
+        Ok(())
+    }
+
+    fn scan_genesis_block<B: Backend>(&mut self, db_tx: &mut StoreTxRw<B>) -> WalletResult<()> {
+        let chain_config = Arc::clone(&self.chain_config);
+        for (output_index, output) in chain_config.genesis_block().utxos().iter().enumerate() {
+            let utxo_source = UtxoSource::Blockchain(BlockHeight::zero());
+            let outpoint = OutPoint::new(
+                OutPointSourceId::BlockReward(self.chain_config.genesis_block_id()),
+                output_index as u32,
+            );
+            self.add_utxo_if_own(db_tx, output, outpoint, utxo_source)?
+        }
+        Ok(())
+    }
+
+    pub fn scan_new_blocks<B: Backend>(
+        &mut self,
+        db_tx: &mut StoreTxRw<B>,
+        mut block_height: BlockHeight,
+        blocks: &[Block],
+    ) -> WalletResult<()> {
+        for block in blocks {
+            let utxo_source = UtxoSource::Blockchain(block_height);
+            let block_id = block.header().block_id().into();
+
+            for (output_index, output) in block.block_reward().outputs().iter().enumerate() {
+                let outpoint =
+                    OutPoint::new(OutPointSourceId::BlockReward(block_id), output_index as u32);
+                self.add_utxo_if_own(db_tx, output, outpoint, utxo_source.clone())?
+            }
+
+            for signed_tx in block.transactions() {
+                let tx_id = signed_tx.transaction().get_id();
+                for input in signed_tx.inputs().iter() {
+                    self.del_utxo_if_own(db_tx, input.outpoint())?
```

Comment:
Sure, but do we need to spend time on this now? I don't see why we would need it in the near future.

Reply:
The problem is that this kind of question can only be answered after suffering. I definitely don't have the foresight to see the consequences of ignoring something that's done for years... but we tend to avoid reinventing the wheel unless we have to and we find a good reason. So, it may be my ignorance playing in here, but I would prefer we do this correctly to avoid problems. 

If you'd like to do this later in the name of getting something working, I don't mind at all. Just please create an issue, write a TODO, and explain the problem at hand.

Reply:
I added TODO.